website/docs/r/nsxt_firewall.html.markdown

```diff
@@ -0,0 +1,117 @@
+
+Provides a resource to manage NSX-T Firewall. Firewalls allow user to control the incoming and 
+outgoing network traffic to and from an NSX-T Data Center Edge Gateway.
+
+## Example Usage 1 (Single rule to allow all IPv4 traffic from anywhere to anywhere)
+```hcl
+resource "vcd_nsxt_firewall" "testing" {
+  org  = "my-org"
+  vdc  = "my-nsxt-vdc"
+
+  edge_gateway_id = data.vcd_nsxt_edgegateway.testing.id
+
+  rule {
+    name        = "allow all IPv4 traffic"
+    direction   = "IN_OUT"
+    ip_protocol = "IPV4"
+  }
+}
+```
+
+## Example Usage 2 (Multiple firewall rules - order matters)
+```hcl
+resource "vcd_nsxt_firewall" "testing" {
+  org  = "my-org"
+  vdc  = "my-nsxt-vdc"
+
+  edge_gateway_id = data.vcd_nsxt_edgegateway.testing.id
+
+  # Rule #1 - Allows in IPv4 traffic from security group `vcd_nsxt_security_group.group1.id`
+  rule {
+    name        = "first rule"
+    direction   = "IN"
+    ip_protocol = "IPV4"
+    source_ids  = [vcd_nsxt_security_group.frontend.id]
+  }
+
+  # Rule #2 - Drops and logs all outgoing IPv6 traffic to `vcd_nsxt_security_group.group.2.id`
+  rule {
+    name            = "drop IPv6 with destination to security group 2"
+    direction       = "OUT"
+    ip_protocol     = "IPV6"
+    destination_ids = [vcd_nsxt_security_group.group2.id]
+    action          = "DROP"
+    logging         = true
+  }
+  
+  # Rule #3 - Allows IPv4 and IPv6 traffic of 2 Application Port Profiles in both directions:
+  # from vcd_nsxt_security_group.group.1.id to all list of security groups vcd_nsxt_security_group.group.*.id
+  # from list of security groups vcd_nsxt_security_group.group.*.id to vcd_nsxt_security_group.group.1.id
+  rule {
+    name            = "test_rule-3"
+    direction       = "IN_OUT"
+    ip_protocol     = "IPV4_IPV6"
+    source_ids      = [vcd_nsxt_security_group.group.1.id]
+    destination_ids = vcd_nsxt_security_group.group.*.id
+    app_port_profile_ids = [app_port_profile_ids = [data.vcd_nsxt_app_port_profile.ssh.id, vcd_nsxt_app_port_profile.custom-app.id]
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `org` - (Optional) The name of organization to use, optional if defined at provider level. Useful
+  when connected as sysadmin working across different organisations.
+* `vdc` - (Optional) The name of VDC to use, optional if defined at provider level.
+* `edge_gateway_id` - (Required) The ID of the edge gateway (NSX-T only). Can be looked up using
+  `vcd_nsxt_edgegateway` datasource
+* `rule` (Required) One or more blocks with [Firewall Rule](#firewall-rule) definitions
+
+<a id="firewall-rule"></a>
+## Firwall Rule
+
+Each Firewall Rule contains following attributes:
+
+* `name` - (Required) Explanatory name for firewall rule (uniqueness not enforced)
+* `direction` - (Required) One of `IN`, `OUT`, or `IN_OUT`
+* `ip_protocol` - (Required) One of `IPV4`,  `IPV6`, or `IPV4_IPV6`
+* `enabled` - (Optional) Defines if the rule is enabled (default `true`)
```

Comment:
Would it be possible to have a test which tries to disable and then enable a rule? Though here it's unlikely, this scenario reminds me one issue in NSX-V which we had long time ago, so a test would be handy.

Reply:
Added and tested.